File: packages/desktop/src/main/nowplaying.ts
import { execFile } from "node:child_process";

export interface NowPlayingInfo {
	title: string;
	artist: string;
	album: string;
	genre: string;
	duration: number; // seconds
	elapsed: number; // seconds
	isPlaying: boolean;
	source: string; // "Music", "Spotify", etc.
	volume: number; // 0–100
}

/** Query the currently playing track from known macOS music apps via osascript */
export async function getNowPlaying(): Promise<NowPlayingInfo | null> {
	// Try each known music app in order
	return (await tryMusic()) ?? (await trySpotify()) ?? null;
}

function tryMusic(): Promise<NowPlayingInfo | null> {
	return queryApp("Music", `
		tell application "System Events"
			if not (exists process "Music") then return "NOT_RUNNING"
		end tell
		tell application "Music"
			if player state is not playing then return "NOT_PLAYING"
			set t to name of current track
			set a to artist of current track
			set al to album of current track
			set g to genre of current track
			set d to duration of current track
			set p to player position
			set v to sound volume
			return t & "||" & a & "||" & al & "||" & g & "||" & d & "||" & p & "||" & v
		end tell
	`, "Music");
}

function trySpotify(): Promise<NowPlayingInfo | null> {
	return queryApp("Spotify", `
		tell application "System Events"
			if not (exists process "Spotify") then return "NOT_RUNNING"
		end tell
		tell application "Spotify"
			if player state is not playing then return "NOT_PLAYING"
			set u to spotify url of current track
			if u starts with "spotify:ad:" then return "NOT_PLAYING"
			set t to name of current track
			set a to artist of current track
			set al to album of current track
			set d to (duration of current track) / 1000
			set p to player position
			set v to sound volume
			return t & "||" & a & "||" & al & "||" & "||" & d & "||" & p & "||" & v
		end tell
	`, "Spotify");
}

function queryApp(
	_appName: string,
	script: string,
	source: string,
): Promise<NowPlayingInfo | null> {
	return new Promise((resolve) => {
		execFile("osascript", ["-e", script], { timeout: 5000 }, (error, stdout) => {
			if (error) {
				resolve(null);
				return;
			}

			const result = stdout.trim();
			if (result === "NOT_RUNNING" || result === "NOT_PLAYING") {
				resolve(null);
				return;
			}

			const parts = result.split("||");
			if (parts.length < 7) {
				resolve(null);
				return;
			}

			resolve({
				title: parts[0],
				artist: parts[1],
				album: parts[2],
				genre: parts[3],
				duration: Number.parseFloat(parts[4]) || 0,
				elapsed: Number.parseFloat(parts[5]) || 0,
				isPlaying: true,
				source,
				volume: Number.parseInt(parts[6], 10) || 0,
			});
		});
	});
}

File: packages/desktop/src-tauri/src/nowplaying.rs
use crate::types::NowPlayingInfo;
use std::process::Command;
use std::time::Duration;
use tokio::time::timeout;

const MUSIC_SCRIPT: &str = r#"
tell application "System Events"
    if not (exists process "Music") then return "NOT_RUNNING"
end tell
tell application "Music"
    if player state is not playing then return "NOT_PLAYING"
    set t to name of current track
    set a to artist of current track
    set al to album of current track
    set g to genre of current track
    set d to duration of current track
    set p to player position
    set v to sound volume
    return t & "||" & a & "||" & al & "||" & g & "||" & d & "||" & p & "||" & v
end tell
"#;

const SPOTIFY_SCRIPT: &str = r#"
tell application "System Events"
    if not (exists process "Spotify") then return "NOT_RUNNING"
end tell
tell application "Spotify"
    if player state is not playing then return "NOT_PLAYING"
    set u to spotify url of current track
    if u starts with "spotify:ad:" then return "NOT_PLAYING"
    set t to name of current track
    set a to artist of current track
    set al to album of current track
    set d to (duration of current track) / 1000
    set p to player position
    set v to sound volume
    return t & "||" & a & "||" & al & "||" & "||" & d & "||" & p & "||" & v
end tell
"#;

pub async fn get_now_playing() -> Option<NowPlayingInfo> {
    if let Some(info) = query_app(MUSIC_SCRIPT, "Music").await {
        return Some(info);
    }
    query_app(SPOTIFY_SCRIPT, "Spotify").await
}

async fn query_app(script: &str, source: &str) -> Option<NowPlayingInfo> {
    let script = script.to_string();
    let source = source.to_string();

    let result = timeout(
        Duration::from_secs(5),
        tokio::task::spawn_blocking(move || {
            Command::new("osascript").args(["-e", &script]).output()
        }),
    )
    .await
    .ok()?
    .ok()?
    .ok()?;

    let stdout = String::from_utf8_lossy(&result.stdout).trim().to_string();

    if stdout == "NOT_RUNNING" || stdout == "NOT_PLAYING" || stdout.is_empty() {
        return None;
    }

    let parts: Vec<&str> = stdout.split("||").collect();
    if parts.len() < 7 {
        return None;
    }

    Some(NowPlayingInfo {
        title: parts[0].to_string(),
        artist: parts[1].to_string(),
        album: parts[2].to_string(),
        genre: parts[3].to_string(),
        duration: parts[4].parse().unwrap_or(0.0),
        elapsed: parts[5].parse().unwrap_or(0.0),
        is_playing: true,
        source,
        volume: parts[6].parse().unwrap_or(0),
    })
}
